src/serial.rs

```diff
@@ -1424,10 +1424,55 @@ macro_rules! halUsartImpl {
                         bb::set(&rcc.$apbXenr, $rcc_bit);
                     }
 
+                    let pclk_freq = clocks.$pclkX().0;
+                    let baud = config.baudrate.0;
+
+                    // The frequency to calculate USARTDIV is this:
+                    //
+                    // (Taken from STM32F411xC/E Reference Manual,
+                    // Section 19.3.4, Equation 1)
+                    //
+                    // 16 bit oversample: OVER8 = 0
+                    // 8 bit oversample:  OVER8 = 1
+                    //
+                    // USARTDIV =          (pclk)
+                    //            ------------------------
+                    //            8 x (2 - OVER8) x (baud)
+                    //
+                    // BUT, the USARTDIV has 4 "fractional" bits, which effectively
+                    // means that we need to "correct" the equation as follows:
+                    //
+                    // USARTDIV =      (pclk) * 16
+                    //            ------------------------
+                    //            8 x (2 - OVER8) x (baud)
+                    //
+                    // When OVER8 is enabled, we can only use the lowest three
+                    // fractional bits, so we'll need to shift those last four bits
+                    // right one bit
+
                     // Calculate correct baudrate divisor on the fly
-                    let div = (clocks.$pclkX().0 + config.baudrate.0 / 2)
-                        / config.baudrate.0;
-                    usart.brr.write(|w| unsafe { w.bits(div) });
+                    let (over8, div) = if (pclk_freq / 16) >= baud {
+                        // We have the ability to oversample to 16 bits, take
+                        // advantage of it.
+                        let div = pclk_freq / baud;
```

Comment:
Not sure why you removed the rounding that was in the original code ?

eg. with a 48 MHz pclk and a 115200 baudrate:
original code: div=417 actual_baudrate=115108
your code: div=416 actual_baudrate=115385

So with your code the actual baudrate is further away from the target baudrate than with the original code.